### PR TITLE
add Metric Unit of measure (UOM) to Nagios output, avoid python crash when running check_ntpmon and no NTP process is found

### PR DESCRIPTION
**Pull request by fbourqui**
The Metric UOM will be dispayed in RRD graph.

### CODE REVIEW

**Comment by fbourqui** [comment]
Hello second commit is to avoid python crash when running check_ntpmon and no NTP process is found, it will return the readable message, and the return code -1 (255), I id not found an elegant way to return(2) :

CRITICAL: No NTP process could be found.  Please check that an NTP server is installed and running.

**Comment by paulgear** [comment]
Hi @fbourqui,

Thanks for the PR - I'm not sure why I didn't see it until just recently.  I'll check my notification settings, but I want to get these changes merged.  I've got a couple of minor comments on the code which I'll add shortly, but if you're not in a position to make changes, then I'll just merge and commit a tidy-up afterwards.

One question: what Nagios setup are you using which makes use of the units of measure, and is there an easy way you can test any updates non-invasively?  I'm spinning up a test Nagios server in a Docker container at the moment, so I should be able to do some basic sanity checking, but I'd like to make sure that your requirements are met by the changes.

Regards,
Paul

File: src/alert.py [hunk 1]
```diff
@@ -79,6 +79,27 @@
     'result': (None, 'd'),
 }
 
+"""
+Metric Unit of measure (UOM) for Nagios
+"""
+_nagios_uom = {
+
+    'frequency': 'ms',
```
Comment (paulgear) [inline src/alert.py:87]: My understanding of frequency is that it's not in milliseconds, but parts-per-million.  I'm not sure if that is something that is easily representable by Nagios' available units of measure; if not, I think microseconds would be more accurate than milliseconds.

File: src/alert.py [hunk 2]
```diff
@@ -170,20 +191,21 @@ def collectmetrics(self, checkobjs, debug):
...
+                pprint.pprint(self.metrics)
+            metrics.addaliases(self.metrics, _aliases)
+            if 'proc' in self.checks:
+                self.checks.append('runtime')
+            if 'trace' in self.checks:
+                self.checks.append('tracehosts')
+                self.checks.append('traceloops')
+                self.checks.append('tracetime')
+            if 'vars' in self.checks and 'offset' not in self.checks:
+                self.checks.append('sysoffset')
```
Comment (paulgear) [inline src/alert.py:208]: It looks like most of this change is just an indent.  If that's the case, I'd prefer to see just:

    if not checkobjs:
        return
    ... # the rest of the function unchanged

File: src/alert.py [hunk 3]
```diff
@@ -307,13 +329,21 @@ def alert_nagios(self, checkobjs, debug):
                 msgs[metric] = self.custom_message(metric, results[metric])
                 if msgs[metric] is None:
                     msgs[metric] = self.mc.message(metric, _formats[metric][0], _formats[metric][1])
+  
+        if not msgs:
+            msgs['no_ntp_proc'] = ('CRITICAL: No NTP process could be found.'
+                '  Please check that an NTP server is installed and running.')
+
```
Comment (paulgear) [inline src/alert.py:336]: Like you say, I think we need to find an elegant method to do this without duplicating the error messages which are already produced under other circumstances.

File: src/alert.py [hunk 4]
```diff
@@ -307,13 +329,21 @@ def alert_nagios(self, checkobjs, debug):
...
             for m in msgs:
                 print(msgs[m])
         else:
             (m, rc) = self.mc.worst_metric(self.checks)
             self.metrics['result'] = self.return_code()
-            print('%s | %s' % (msgs[m], self.report()))
+            if 'no_ntp_proc' in msgs:
+                print(msgs['no_ntp_proc'])
+            else:
+                print('%s | %s' % (msgs[m], self.report()))
```
Comment (paulgear) [inline src/alert.py:346]: Same comment as above applies here.

**Review by paulgear — CHANGES_REQUESTED** [review]
@fbourqui After your comment about finding an elegant way of producing the correct output, I had another look at the existing master, and I've got a change prepared which is much shorter and hopefully clearer.  I think we'd be best to roll back your PR to being just the Units of Measure change, and I'll put my proposed change up in a separate branch for your testing & feedback.

**Comment by fbourqui** [comment]
Hello Paul,





I'm using your nagios plugin with librenms (yes it has support for nagios plugin).



I'm quite busy right now with other topics, but I will take some time to check if your changes are working for my librenms use case, I will let you know when I have time to validate that, most probably next week.



Cheers,

Fred





---- On Mon, 05 Oct 2020 10:37:28 +0200 Paul Gear <notifications@github.com> wrote ----





Hi https://github.com/fbourqui,

Thanks for the PR - I'm not sure why I didn't see it until just recently.  I'll check my notification settings, but I want to get these changes merged.  I've got a couple of minor comments on the code which I'll add shortly, but if you're not in a position to make changes, then I'll just merge and commit a tidy-up afterwards.

One question: what Nagios setup are you using which makes use of the units of measure, and is there an easy way you can test any updates non-invasively?  I'm spinning up a test Nagios server in a Docker container at the moment, so I should be able to do some basic sanity checking, but I'd like to make sure that your requirements are met by the changes.

Regards,
 Paul

**Comment by paulgear** [comment]
Hi @fbourqui, the main branch has moved along a bit since this PR and I've merged the change which correctly detects the non-running NTP service.  If you have any further changes to propose around the units of measure issue, please drop me a new PR forked from the current main branch.

**Pull request closed without merging**